File: packages/sunrey-chain/rust/crates/protocol/src/commitments.rs
use crate::block::BlockHeader;
use crate::codec::{encode_bytes, encode_string};
use crate::genesis::GenesisV1;
use crate::transaction::UnsignedTransaction;
use crate::{
    DOMAIN_BLOCK_ID, DOMAIN_GENESIS, DOMAIN_LEAF, DOMAIN_MERKLE, DOMAIN_SIG, DOMAIN_STATE_ROOT,
    DOMAIN_TX_ID, DOMAIN_TX_ROOT,
};

pub const HASH_SIZE: usize = 32;
pub type Hash32 = [u8; HASH_SIZE];

/// Hash provider port. Business modules must not hard-code an algorithm.
pub trait DomainHasher {
    fn hash(&self, domain: &str, payload: &[u8]) -> Hash32;
}

pub fn domain_payload(domain: &str, payload: &[u8]) -> Vec<u8> {
    let mut out = Vec::new();
    encode_string(&mut out, domain);
    out.extend_from_slice(payload);
    out
}

pub fn transaction_id(hasher: &dyn DomainHasher, unsigned: &UnsignedTransaction) -> Hash32 {
    hasher.hash(DOMAIN_TX_ID, &unsigned.encode())
}

pub fn unsigned_signature_payload(
    hasher: &dyn DomainHasher,
    unsigned: &UnsignedTransaction,
) -> Hash32 {
    hasher.hash(DOMAIN_SIG, &unsigned.encode())
}

pub fn genesis_hash(hasher: &dyn DomainHasher, genesis: &GenesisV1) -> Hash32 {
    hasher.hash(DOMAIN_GENESIS, &genesis.encode())
}

pub fn block_id(hasher: &dyn DomainHasher, header: &BlockHeader) -> Hash32 {
    hasher.hash(DOMAIN_BLOCK_ID, &header.encode())
}

pub fn merkle_root(hasher: &dyn DomainHasher, domain: &str, leaves: &[Hash32]) -> Hash32 {
    if leaves.is_empty() {
        return hasher.hash(domain, &[]);
    }
    let mut layer: Vec<Hash32> = leaves.to_vec();
    while layer.len() > 1 {
        if layer.len() % 2 == 1 {
            if let Some(last) = layer.last().copied() {
                layer.push(last);
            }
        }
        let mut next = Vec::with_capacity(layer.len() / 2);
        for pair in layer.chunks_exact(2) {
            let mut payload = Vec::with_capacity(HASH_SIZE * 2);
            payload.extend_from_slice(&pair[0]);
            payload.extend_from_slice(&pair[1]);
            next.push(hasher.hash(DOMAIN_MERKLE, &payload));
        }
        layer = next;
    }
    layer[0]
}

pub fn transaction_root(hasher: &dyn DomainHasher, tx_ids: &[Hash32]) -> Hash32 {
    merkle_root(hasher, DOMAIN_TX_ROOT, tx_ids)
}

pub fn state_root(hasher: &dyn DomainHasher, entries: &[(Vec<u8>, Vec<u8>)]) -> Hash32 {
    let mut leaves = Vec::with_capacity(entries.len());
    for (key, value) in entries {
        let mut payload = Vec::new();
        encode_bytes(&mut payload, key);
        encode_bytes(&mut payload, value);
        leaves.push(hasher.hash(DOMAIN_LEAF, &payload));
    }
    merkle_root(hasher, DOMAIN_STATE_ROOT, &leaves)
}

File: packages/sunrey-chain/src/blocks/commitments.ts
// @ts-nocheck
/**
 * Deterministic Merkle commitments for blocks and monetary state.
 *
 * Domains and tree rules match `packages/sunrey-chain/rust/crates/protocol/src/commitments.rs`.
 */

import { createHash } from 'node:crypto';

import { domainPayload, encodeBytes, encodeString, encodeU64 } from './codec.ts';
import type { CanonicalBlockHeader } from './types.ts';

export const HASH_SIZE = 32 as const;

export const DOMAIN_BLOCK_ID = 'sunrey.blockid.v1' as const;
export const DOMAIN_MERKLE = 'sunrey.merkle.v1' as const;
export const DOMAIN_LEAF = 'sunrey.leaf.v1' as const;
export const DOMAIN_STATE_ROOT = 'sunrey.stateroot.v1' as const;
export const DOMAIN_TX_ROOT = 'sunrey.txroot.v1' as const;

export type Hash32 = Uint8Array;

export function hashDomain(domain: string, payload: Uint8Array): Hash32 {
  return createHash('sha256').update(domainPayload(domain, payload)).digest();
}

export function hashToHex(hash: Hash32): string {
  return Buffer.from(hash).toString('hex');
}

export function hashFromHex(text: string): Hash32 {
  if (!/^[0-9a-f]{64}$/i.test(text)) {
    throw new TypeError('hash must be 32-byte hex');
  }
  return Buffer.from(text, 'hex');
}

export function merkleRoot(domain: string, leaves: readonly Hash32[]): Hash32 {
  if (leaves.length === 0) {
    return hashDomain(domain, new Uint8Array());
  }
  let layer = leaves.map((leaf) => Buffer.from(leaf));
  while (layer.length > 1) {
    if (layer.length % 2 === 1) {
      const last = layer[layer.length - 1];
      if (last) {
        layer.push(Buffer.from(last));
      }
    }
    const next: Buffer[] = [];
    for (let index = 0; index < layer.length; index += 2) {
      const left = layer[index];
      const right = layer[index + 1];
      if (!left || !right) {
        throw new Error('invalid merkle layer');
      }
      next.push(hashDomain(DOMAIN_MERKLE, Buffer.concat([left, right])));
    }
    layer = next;
  }
  const root = layer[0];
  if (!root) {
    throw new Error('empty merkle root');
  }
  return root;
}

export function transactionRoot(txIds: readonly Hash32[]): Hash32 {
  return merkleRoot(DOMAIN_TX_ROOT, txIds);
}

export function stateLeaf(key: Uint8Array, value: Uint8Array): Hash32 {
  const payload = Buffer.concat([encodeBytes(key), encodeBytes(value)]);
  return hashDomain(DOMAIN_LEAF, payload);
}

export function stateRoot(entries: ReadonlyMap<string, Uint8Array> | readonly [string, Uint8Array][]): Hash32 {
  const pairs = entries instanceof Map ? [...entries.entries()] : [...entries];
  pairs.sort(([left], [right]) => (left < right ? -1 : left > right ? 1 : 0));
  const leaves = pairs.map(([key, value]) => stateLeaf(Buffer.from(key, 'utf8'), value));
  return merkleRoot(DOMAIN_STATE_ROOT, leaves);
}

export function encodeBlockHeader(header: CanonicalBlockHeader): Buffer {
  return Buffer.concat([
    encodeString('BlockHeaderV1'),
    encodeU32FromNumber(header.version),
    encodeString(header.networkId),
    encodeString(header.chainId),
    encodeU64(header.height),
    encodeBytes(header.parentBlockHash),
    encodeBytes(header.transactionRoot),
    encodeBytes(header.previousStateCommitment),
    encodeBytes(header.resultingStateCommitment),
    encodeBytes(header.validatorSetHash),
    encodeBytes(header.consensusParameterHash),
    encodeString(header.protocolVersion),
    encodeBytes(header.moduleRegistryHash),
    encodeBytes(header.codecRegistryHash),
    encodeBytes(header.cryptoPolicyHash),
    encodeU64(header.timestampUnixMs),
    encodeString(header.proposer),
    encodeString(header.cryptoSuiteId),
    encodeU32FromNumber(header.round),
    encodeBytes(header.consensusCertificateHash),
    encodeExtensionCommitments(header.extensionCommitments),
  ]);
}

function encodeU32FromNumber(value: number): Buffer {
  const out = Buffer.alloc(4);
  out.writeUInt32BE(value >>> 0);
  return out;
}

function encodeExtensionCommitments(extensions: Readonly<Record<string, Hash32>>): Buffer {
  const keys = Object.keys(extensions).sort();
  const parts = [encodeU32FromNumber(keys.length)];
  for (const key of keys) {
    const value = extensions[key];
    if (!value) {
      continue;
    }
    parts.push(encodeString(key), encodeBytes(value));
  }
  return Buffer.concat(parts);
}

export function blockId(header: CanonicalBlockHeader): Hash32 {
  return hashDomain(DOMAIN_BLOCK_ID, encodeBlockHeader(header));
}

export function transactionIdFromBytes(canonicalBytes: Uint8Array): Hash32 {
  return hashDomain('sunrey.txid.v1', canonicalBytes);
}
